streams/shared/src/main/scala/zio/stream/ZChannel.scala

```diff
@@ -1938,115 +1940,131 @@ object ZChannel {
+        val permits        = Semaphore.unsafe.make(n0)(Unsafe)
+
+        type Pull[A] = ZIO[Env, Either[OutErr, OutDone], A]
+
+        def evaluatePull(pull: Pull[OutElem]): URIO[Env, Unit] =
+          pull
+            .flatMap(outElem => outgoing.offer(Result.Value(outElem)))
+            .forever
+            .catchAllCause(cause =>
+              cause.failureOrCause match {
+                case Left(x: Right[OutErr, OutDone]) =>
+                  ZIO.succeed {
+                    lastDone.updateAndGet {
+                      case null     => x.value
+                      case lastDone => f(lastDone, x.value)
+                    }
+                  }
+                case Left(l: Left[OutErr, OutDone]) =>
+                  outgoing.offer(Result.Error(l.value)) *> errorSignal.succeed(()).unit
+                case Right(cause) =>
+                  outgoing.offer(Result.Fatal(cause)) *> errorSignal.succeed(()).unit
+              }
             )
+
+        type Channel = ZChannel[Env, InErr, InElem, InDone, OutErr, OutElem, OutDone]
+
+        def pullStrategy(childScope: Scope, pull: Pull[Channel]): Pull[Unit] =
+          mergeStrategy0 match {
+            case MergeStrategy.BackPressure =>
+              pull.flatMap { channel =>
+                val latch = Promise.unsafe.make[Nothing, Unit](fiberId)(Unsafe)
+                val raceIOs =
+                  ZIO.scopedWith { scope =>
+                    (incoming >>> channel)
+                      .toPullInAlt(scope)
+                      .flatMap(evaluatePull)
+                  }
+
+                permits
+                  .withPermit(latch.succeed(()) *> raceIOs)
+                  .interruptible
+                  .forkIn(childScope) *> latch.await
+              }
+            case MergeStrategy.BufferSliding =>
+              pull.flatMap { channel =>
+                val canceler = Promise.unsafe.make[Nothing, Unit](fiberId)(Unsafe)
+                val latch    = Promise.unsafe.make[Nothing, Unit](fiberId)(Unsafe)
+
+                for {
+                  size <- cancelers.size
+                  _    <- ZIO.when(size >= n0)(cancelers.take.flatMap(_.succeed(())))
+                  _    <- cancelers.offer(canceler)
+                  raceIOs =
+                    ZIO.scopedWith { scope =>
+                      (incoming >>> channel)
+                        .toPullInAlt(scope)
+                        .flatMap(evaluatePull(_).race(canceler.await.interruptible))
+                    }
+                  _ <- permits
+                         .withPermit(latch.succeed(()) *> raceIOs)
+                         .interruptible
+                         .forkIn(childScope)
+                  _ <- latch.await
+                } yield ()
+              }
           }
 
-        consumer.embedInput(input)
+        for {
+          _          <- scope.addFinalizer(outgoing.shutdown)
+          _          <- scope.addFinalizer(cancelers.shutdown)
```

Comment:
@kyri-petrou I wonder if we shouldn't wrap these 2 calls in an `ZIO.uninterruptible` 🤔